check_license_go_code.sh

```diff
@@ -13,33 +43,85 @@ cat > license.tmp <<EOF
 # // Copyright <copyright_year> Northern.tech AS
 # //
-lines=$(($lines + 2))
+LINES_OS=$(($LINES_OS + 2))
 
-ret=0
-for each in $(find . -type f \( ! -regex '.*/\..*' ! -path "./Godeps/*" ! -path "./vendor/*" -name '*.go' \)); do
-  modified_year=$(git log --follow -n1 --format=%ad --date=format:%Y -- $each)
-  
-  echo "Checking $each for correct license header; last modified in $modified_year"
-
-  head -n $lines $each | tail -n +3 | diff -qu license.tmp - > /dev/null
-  if [ ! "$?" -eq "0" ]; then
-    echo "!!! FAILED license check on $each"
-    ret=1
-  else
-    copyright_modified=$(echo "// Copyright <copyright_year> Northern.tech AS" | sed "s/<copyright_year>/$modified_year/g")
-    copyright_file=$(head -n 1 $each)
-    if [ "$copyright_modified" == "$copyright_file" ]; then
-      echo "License check passed on $each"
+cat > license-ent.tmp <<EOF
+//    All Rights Reserved
+EOF
+LINES_ENT=$(cat license-ent.tmp | wc -l)
+# we need to add two extra lines missing from the license preamble
+# // Copyright <copyright_year> Northern.tech AS
+# //
+LINES_ENT=$(($LINES_ENT + 2))
+
+TEST_RESULT=0
+LATEST_OS_COMMIT=
+
+check_file() {
+    local file="$1"
+
+    local license
+    local lines
+    local lic_type
+    if is_enterprise "$file"; then
+        license="license-ent.tmp"
+        lines=$LINES_ENT
+        lic_type="Enterprise"
     else
-      echo "!!! FAILED license check on $each; make sure copyright year matches last modified year of the file"
-      ret=1
+        license="license-os.tmp"
+        lines=$LINES_OS
+        lic_type="Open Source"
     fi
-  fi
+
+    modified_year=$(git log --follow -n1 --format=%ad --date=format:%Y -- "$file")
+
+    head -n $lines "$file" | tail -n +3 | diff -qu "$license" - > /dev/null
+    if [ ! "$?" -eq "0" ]; then
+        echo "!!! FAILED license check on $file. Expected this $lic_type license:"
+        cat "$license"
+        TEST_RESULT=1
+    else
+        copyright_modified=$(echo "// Copyright <copyright_year> Northern.tech AS" | sed "s/<copyright_year>/$modified_year/g")
+        copyright_file="$(head -n 1 "$file")"
+        if [ "$copyright_modified" != "$copyright_file" ]; then
+            echo "!!! FAILED license check on $file; make sure copyright year matches last modified year of the file ($modified_year)"
+            TEST_RESULT=1
+        fi
+    fi
+}
+
+is_enterprise() {
+    local file="$1"
+
+    if [ -z "$ENT_COMMIT" ]; then
+        # If there is no Enterprise commit specified, then this isn't an
+        # Enterprise repository, so everything is Open Source.
+        return 1
+    fi
+
+    # Find the latest commit that is not a descendant of the Enterprise
+    # commit. This should be the latest Open Source commit. This doesn't change
+    # over the course of a run, so cache it.
+    if [ -z "$LATEST_OS_COMMIT" ]; then
+        LATEST_OS_COMMIT=$(git rev-list $ENT_COMMIT..HEAD --ancestry-path --boundary --date-order | grep "^-" | head -n1 | grep -o '[0-9a-f]*')
```

Comment:
It's always nice to learn something new from git :)

Just curious: why do we need the last `grep -o`?

Reply:
Yes, it took me some time to figure that one out!

The final `grep -o` is needed because the second last grep output is prefixed with a `-`, and we need to get rid of it with `-o`.

Reply:
Yep, I see it now. Thanks.